### DISCUSSION TITLE
Plot many 3D lines with specified colors

### DISCUSSION THREAD

**Discussion by qtn-mtp** [pyvista/pyvista#3570] [Q&A] [answered]
Hello, 
I would like to be able to efficiently plot many independent 3D lines with specific colors (without loops). I have searched in last few issues and read [this](https://github.com/pyvista/pyvista-support/issues/359) or [that](https://github.com/pyvista/pyvista/discussions/2387) but it doesn't solve my problem.
 
The only way I have found is to create lines with Polydata as in the function [line_segments_from_points](https://docs.pyvista.org/api/core/helpers.html#line-segments-from-points) (code source [at line 536](https://github.com/pyvista/pyvista/blob/c8017a55d63c7a5eef57e43a728007623981ae9a/pyvista/utilities/helpers.py#L536) in file [/pyvista/utilities/helpers.py](https://github.com/pyvista/pyvista/blob/c8017a55d63c7a5eef57e43a728007623981ae9a/pyvista/utilities/helpers.py) ) and use parameters cmap and scalars in [add_mesh](https://docs.pyvista.org/api/plotting/_autosummary/pyvista.Plotter.add_mesh.html#add-mesh) to specify independently a color to each line. 

Here is my code :
```python
import numpy as np
import pyvista as pv
plotter = pv.Plotter()

pts_1 = np.array([(1,1,0),(2,2,0),  (3,3,0), (4,4,0),   (5,5,0)])
pts_2 = np.array([(1,1,3),(2,2,3.5),(3,3,4), (4,4,4.5), (5,5,5)])
nb_lines = len(pts_1)
pts = np.r_[(pts_1, pts_2)]
my_lines = pv.PolyData()
my_lines.points = pts
cells = np.full((nb_lines, 3), 2, dtype=np.int_)
cells[:, 1] = np.arange(0, nb_lines, dtype=np.int_)
cells[:, 2] = np.arange(nb_lines, nb_lines*2, dtype=np.int_)
my_lines.lines = cells    

colormaps = ['green', 'red', 'blue', 'orange', 'green']
colors = np.arange(nb_lines)

plotter.add_mesh(my_lines, line_width=10, render_lines_as_tubes=True, 
                  scalars=colors, cmap=colormaps)
```
![image](https://user-images.githubusercontent.com/109147535/200598647-9e5dc835-acbc-4055-9830-e95bac255107.png)

It works but is there a way to do the same thing without using cmap ?

If it is not possible, my problem with that is I can't specify colors by numbers in cmap. This is because when we override cmap with a custom color map, it only accepts a list of color strings.
From documentation of add_Mesh : "You can also specify a list of colors to override an existing colormap with a custom one.  For example, to  create a three color colormap you might specify `['green', 'red', 'blue']``."
You can't for example do this :
```python
plotter.add_mesh(my_lines, line_width=10, render_lines_as_tubes=True, 
                  scalars=colors, cmap=[(0,1,0), (1,0,0), (0,0,1), (0,1,1), (0,1,0)])
```
It is due to a "raise TypeError" which is for me useless for my case in function [get_cmap_safe](https://github.com/pyvista/pyvista/blob/c8017a55d63c7a5eef57e43a728007623981ae9a/pyvista/plotting/colors.py#L984) :
```python
def get_cmap_safe(cmap):
    """Fetch a colormap by name from matplotlib, colorcet, or cmocean."""
    # ...
    elif isinstance(cmap, list):
        for item in cmap:
            if not isinstance(item, str):
                raise TypeError('When inputting a list as a cmap, each item should be a string.')
        from matplotlib.colors import ListedColormap
        cmap = ListedColormap(cmap)

    return cmap
```
If I comment out exception, it works as I want it to, since [matplotlib.colors.ListedColormap](https://matplotlib.org/stable/api/_as_gen/matplotlib.colors.ListedColormap.html#matplotlib-colors-listedcolormap) take into account list of color tuples.
So, maybe it will be nice to add this case for cmap. Maybe, my message is more suitable in request feature ?
Thanks in advance
Sincerly

**Comment by banesullivan** [comment]
Colormapping is for mapping a scalar field specifically. If you want to color each line with a specified, solid color, you'll want to split the mesh by each line and add them to the scene individually while specifying the `color`.

The mesh can be split with the `split_bodies()` filter: https://docs.pyvista.org/api/core/_autosummary/pyvista.DataSet.split_bodies.html#split-bodies

(there are other ways to split based on cells, but `split_bodies` is probably the simplest)

```py
lines = my_lines.split_bodies()
colors = [...]  # some list with length `len(lines)`

for i, line in enumerate(lines):
    plotter.add_mesh(line, color=colors[i], ...)
```

Hope this helps

**Reply by qtn-mtp** [reply]
Thank you for your answer.
I agree with you but what you proposed is done with loop. Your solution would be good in C/C++ but in python it is very slow, and in my case I have many lines.
Furthermore, I suppose a lot of mesh slows down the 3D viewer. With my solution, I have only one mesh for unlimited number of lines

**Reply by banesullivan** [reply]
You may need to use the composite plotting features then. Ref https://docs.pyvista.org/api/plotting/_autosummary/pyvista.Plotter.add_composite.html#add-composite

```py
lines = my_lines.split_bodies()
colors = [...]  # some list with length `len(lines)`

plotter.add_composite(lines, color=colors, ...)
```

(this is psuedo code... not sure if colors will work that way)

cc @akaszynski as he's done a lot more work with composite rendering like this

**Reply by qtn-mtp** [reply]
I'm going to test this, thank you !

**Reply by qtn-mtp** [reply]
I tested and the problem remains since split_bodies() use a loop.
Thank you for your help but for the moment I am going to keep my solution even if I dont like it since I have to comment the exception in get_cmap_safe().
Maybe you should modify get_cmap_safe() to take into account list of color tuples as matplotlib. It can be useful even in other cases when you need to use matplotlib.colors.ListedColormap in add_mesh.

**Reply by banesullivan** [reply]
Ah, my bad. Didn't realize that had an internal for loop. Also how big are your data? Can you verify how much that for-loop slows things down (just curious how big of a mesh you are dealing with)?

> Maybe you should modify get_cmap_safe() to take into account list of color tuples as matplotlib....

We currently support [`ListedColormap`s from matplotlib](https://matplotlib.org/stable/api/_as_gen/matplotlib.colors.ListedColormap.html#matplotlib.colors.ListedColormap). See some of the examples [here](https://docs.pyvista.org/examples/02-plot/cmap.html)

Though your original question was hoping not to use the `cmap` or any scalar color mapping.

The trick is going to be splitting your lines into individual meshes so that you can use the `color` argument. Perhaps there is a way for us to improve the split_bodies filter for large data?.. we'll have to look into this

**Reply by qtn-mtp** [reply]
> We currently support [`ListedColormap`s from matplotlib](https://matplotlib.org/stable/api/_as_gen/matplotlib.colors.ListedColormap.html#matplotlib.colors.ListedColormap). See some of the examples [here](https://docs.pyvista.org/examples/02-plot/cmap.html)

Yes but for example, you can do this :
```python
mesh.plot(scalars=scalars, cmap=['red', 'green','blue'])
```
but because of get_cmap_safe() you can't do this (which is normally possible in matplotlib):
```python
mesh.plot(scalars=scalars, cmap=[(1,0,0), (0,1,0),(0,0,1)])
```
To solve the problem with get_cmap_safe, I modity it by commenting the exception with a monkey patching which is horrible, but at least, I don't modify source code of pyvista (only in runtime) :
```python
from pyvista.utilities.misc import has_module
import pyvista as pv
IPYGANY_MAP = {
    'reds': 'Reds',
    'spectral': 'Spectral',
}
def my_get_cmap_safe(cmap):
    """Fetch a colormap by name from matplotlib, colorcet, or cmocean."""
    if isinstance(cmap, str):
        # check if this colormap has been mapped between ipygany
        if cmap in IPYGANY_MAP:
            cmap = IPYGANY_MAP[cmap]

        # Try colorcet first
        if has_module('colorcet'):
            import colorcet

            try:
                return colorcet.cm[cmap]
            except KeyError:
                pass

        # Try cmocean second
        if has_module('cmocean'):
            import cmocean

            try:
                return getattr(cmocean.cm, cmap)
            except AttributeError:
                pass

        # Else use Matplotlib
        if not has_module('matplotlib'):
            raise ImportError(
                'The use of custom colormaps requires the installation of matplotlib.'
            )  # pragma: no cover

        from matplotlib import colormaps, colors

        if not isinstance(cmap, colors.Colormap):
            cmap = colormaps[cmap]

    elif isinstance(cmap, list):
        # for item in cmap:
        #     if not isinstance(item, str):
        #         raise TypeError('When inputting a list as a cmap, each item should be a string.')

        if not has_module('matplotlib'):
            raise ImportError(
                'The use of custom colormaps requires the installation of matplotlib.'
            )  # pragma: no cover

        from matplotlib.colors import ListedColormap

        cmap = ListedColormap(cmap)

    return cmap
pv.plotting.get_cmap_safe.__code__ = my_get_cmap_safe.__code__
```

**Reply by qtn-mtp** [reply]
> Can you verify how much that for-loop slows things down (just curious how big of a mesh you are dealing with)?

Here is a minimal example code to compare the "loop method" with the "cmap method". I test to plot 100 simple lines for 100 iterations in this example :
```python
import numpy as np
import pyvista as pv
import time
from pyvista.utilities.misc import has_module

IPYGANY_MAP = {'reds': 'Reds', 'spectral': 'Spectral',}
def my_get_cmap_safe(cmap):
    """Fetch a colormap by name from matplotlib, colorcet, or cmocean."""
    if isinstance(cmap, str):
        # check if this colormap has been mapped between ipygany
        if cmap in IPYGANY_MAP:
            cmap = IPYGANY_MAP[cmap]
        # Try colorcet first
        if has_module('colorcet'):
            import colorcet
            try:
                return colorcet.cm[cmap]
            except KeyError:
                pass
        # Try cmocean second
        if has_module('cmocean'):
            import cmocean
            try:
                return getattr(cmocean.cm, cmap)
            except AttributeError:
                pass
        # Else use Matplotlib
        if not has_module('matplotlib'):
            raise ImportError(
                'The use of custom colormaps requires the installation of matplotlib.'
            )  # pragma: no cover
        from matplotlib import colormaps, colors
        if not isinstance(cmap, colors.Colormap):
            cmap = colormaps[cmap]
    elif isinstance(cmap, list):
#####   I commented the 3 following lines   #####
        # for item in cmap:
        #     if not isinstance(item, str):
        #         raise TypeError('When inputting a list as a cmap, each item should be a string.')
        if not has_module('matplotlib'):
            raise ImportError(
                'The use of custom colormaps requires the installation of matplotlib.'
            )  # pragma: no cover
        from matplotlib.colors import ListedColormap
        cmap = ListedColormap(cmap)
    return cmap
pv.plotting.get_cmap_safe.__code__ = my_get_cmap_safe.__code__

def lines(pts_1, pts_2):
    nb_lines = pts_1.shape[0]
    points = np.r_[(pts_1, pts_2)]
    poly = pv.PolyData()
    poly.points = points
    cells = np.full((nb_lines, 3), 2, dtype=np.int_)
    cells[:, 1] = np.arange(0, nb_lines, dtype=np.int_)
    cells[:, 2] = np.arange(nb_lines, nb_lines*2, dtype=np.int_)
    poly.lines = cells
    return poly

if __name__=='__main__':
    nb_lines = 100
    theta = np.radians(np.linspace(0,360*5,nb_lines))
    r = np.linspace(0,1,nb_lines)
    x_2 = r*np.cos(theta)
    y_2 = r*np.sin(theta)
    h = np.linspace(5,0.5,nb_lines)
    pts_1 = np.c_[x_2, y_2, np.zeros(nb_lines)]
    pts_2 = np.c_[x_2, y_2, np.ones(nb_lines)*h]
    colors = list(np.random.random((nb_lines,3)))

    nb_iteration = 100
    vec_t0 = np.zeros(nb_iteration)
    vec_t1 = np.zeros(nb_iteration)
    for method in (0,1):
        for i in range(nb_iteration):
            plotter = pv.Plotter()
            if i > 0:
                t = time.time()
            if method == 0: # "Loop method"
                my_lines = lines(pts_1, pts_2)
                my_lines = my_lines.split_bodies()
                for j, line in enumerate(my_lines):
                    plotter.add_mesh(line, color=colors[j])
            else: # "cmap method"
                my_lines = lines(pts_1, pts_2)
                plotter.add_mesh(my_lines, scalars=np.arange(nb_lines), cmap=list(colors),
                                 show_scalar_bar=False)
            if i == 0:
                plotter.show()
            elif method == 0:
                vec_t0[i] = time.time() - t
                print(f"method 0, iteration {i} : {vec_t0[i]}")
            elif method == 1:
                vec_t1[i] = time.time() - t
                print(f"method 1, iteration {i} : {vec_t1[i]}")
            plotter.close()

    print(f"\nNumber of lines : {nb_lines}")
    print(f"Number of iterations : {nb_iteration}")

    print(f"\nMethod with loop")
    print(f"Mean time per iteration : {np.sum(vec_t0)/nb_iteration}")
    print(f"Total time : {np.sum(vec_t0)}\n")

    print(f"Method with cmap trick")
    print(f"Mean time per iteration : {np.sum(vec_t1)/nb_iteration}")
    print(f"Total time : {np.sum(vec_t1)}\n")
```

Here is the time I measured, so you can observe the big difference :
```
Number of lines : 100
Number of iterations : 100

Method with loop
Mean time per iteration : 0.2295643925666809
Total time : 22.95643925666809

Method with cmap trick
Mean time per iteration : 0.0036811351776123045
Total time : 0.36811351776123047
```

**Comment by MatthewFlamm** [comment, marked as answer]
I have not read the other thread in detail, so sorry if this was addressed or there is some limitation.  This seems much simpler to me, and I think it does what you are looking for:

```py
import numpy as np
import pyvista as pv
plotter = pv.Plotter()

pts_1 = np.array([(1,1,0),(2,2,0),  (3,3,0), (4,4,0),   (5,5,0)])
pts_2 = np.array([(1,1,3),(2,2,3.5),(3,3,4), (4,4,4.5), (5,5,5)])
nb_lines = len(pts_1)
pts = np.r_[(pts_1, pts_2)]
my_lines = pv.PolyData()
my_lines.points = pts
cells = np.full((nb_lines, 3), 2, dtype=np.int_)
cells[:, 1] = np.arange(0, nb_lines, dtype=np.int_)
cells[:, 2] = np.arange(nb_lines, nb_lines*2, dtype=np.int_)
my_lines.lines = cells    

rgb = [[0, 255, 0], [255, 0, 0], [0, 0, 255], [0, 0, 255], [255, 255, 0], [0, 255, 0]]
my_lines["colors"] = rgb  # set as cell data, could also provide directly as scalars in `add_mesh` call.

plotter.add_mesh(my_lines, line_width=10, render_lines_as_tubes=True, 
                  scalars="colors", rgb=True)
```
![image](https://user-images.githubusercontent.com/39341281/201388495-96c86ff2-cdf5-4dd6-aa09-b5ca3457dd45.png)

**Reply by qtn-mtp** [reply]
Thank you very much ! Indeed, it's exactly what I'm looking for. 
That produces the same result as mine without my horrible monkey patch and the use of cmap. I didn't know the effect of rgb parameters in add_mesh.
The best solution !
